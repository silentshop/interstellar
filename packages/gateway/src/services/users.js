import stellar from 'stellar-sdk';

class Users {

    static verifySignature(publicKey, payload, signature){
        return stellar.Keypair.fromPublicKey(publicKey).verify(payload, Buffer.from(signature, 'base64'));
    }

}

export default Users;
