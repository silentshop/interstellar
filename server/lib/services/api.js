"use strict";
// https://coursetro.com/posts/code/84/Setting-up-an-Angular-4-MEAN-Stack-(Tutorial)
/** express */
import { Router } from "express";
const router = Router();
import userUtils from "./utils.js";
/** Express */
import expressImport from "express";
const expressEngine = expressImport();
/** Cors */
import cors from "cors";
expressEngine.use(cors({ origin: true }));
/** Import Routes */
import cartRoute from "./cart.js";
import categoryRoute from "./category.js";
import chatRoute from "./chat.js";
import usersRoute from "./users.js";
import productsRoute from "./products.js";
/** Assign Routes */
router.use("/categories", categoryRoute);
router.use("/cart", cartRoute);
router.use("/chat", chatRoute);
router.use("/products", productsRoute);
router.use("/users", usersRoute);
/** Exports */
const exporter = {
    router: router,
    expressEngine: expressEngine
};
export default exporter;
//# sourceMappingURL=api.js.map
