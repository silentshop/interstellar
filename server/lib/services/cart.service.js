const firedb = require("../../_firebase");

class CartService {

    constructor() {
        this.cartItemsCollection = firedb.collection('user-cart');
    }

    //
    // ────────────────────────────────────────────────────────────────────────────── I ──────────
    //   :::::: P U B L I C   C R U D   M E T H O D S : :  :   :    :     :        :          :
    // ────────────────────────────────────────────────────────────────────────────────────────
    //
    // todo: pass context ID separately
    addToCart(userID, cartItem) {
        const doc = this.cartItemsCollection.doc(cartItem.userID).collection("cartItems").doc();
        const docID = doc.id;
        cartItem.id = docID;
        return doc
            .set(cartItem)
            .then((documentSnapshot) => docID);
    }

    getMyCart(userID) {
        console.log(userID)
        return this.cartItemsCollection
            .doc(userID)
            .collection("cartItems")
            .get()
            .then(snapshot => 
                snapshot.docs.map((docSnapshot) => 
                    docSnapshot.data()
                )
            );
    }

    updateCartItem(cartItem) {
        return this.cartItemsCollection
                .doc(cartItem.id)
                .update(cartItem, { merge: true})
                .then((documentSnapshot) => documentSnapshot);
    }

    removeFromCart(userID, cartItemID) {
        return this.cartItemsCollection
            .doc(userID)
            .collection("cartItems")
            .doc(cartItemID)
            .delete()
            .then((documentSnapshot) => 
                documentSnapshot
            );

    }
    
    // todo: batching 
    emptyCart(userID) {
            // return this,this.cartItemsCollection
            //     .doc(cartItemID)
            //     .delete()
            //     .then((documentSnapshot) => userID);
    }
    
    // ────────────────────────────────────────────────────────────────────────────────
}

module.exports = new CartService();
