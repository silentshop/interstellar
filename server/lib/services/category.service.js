const firedb = require("../../_firebase");

class CategoryService {

    constructor() {
        this.categorysCollection = firedb.collection('categories');
    }

    //
    // ────────────────────────────────────────────────────────────────────────────── I ──────────
    //   :::::: P U B L I C   C R U D   M E T H O D S : :  :   :    :     :        :          :
    // ────────────────────────────────────────────────────────────────────────────────────────
    //
    createNewCategory(category) {
        const doc = this.categorysCollection.doc();
        const docID = doc.id;
        category.id = docID;
        console.log(category)
        return doc
            .set(category)
            .then((documentSnapshot) => docID);
    }

    getAllCategories() {
        return this.categorysCollection.get().then(snapshot => snapshot.docs.map((docSnapshot) => docSnapshot.data()));
    }

    /**
     * @param  {string} categoryID
     */
    getProductsByCategory(category) {
        console.log(category)
        return this.categorysCollection
            .doc(category)
            .collection('products')
            .get()
            .then(snapshot => 
                snapshot.docs.map((docSnapshot) => 
                    docSnapshot.data()
                )
            );
    }
    // ────────────────────────────────────────────────────────────────────────────────
}

CategoryService = new CategoryService();
module.exports = CategoryService;
