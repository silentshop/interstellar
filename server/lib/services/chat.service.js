// todo: duplication
import firedb  from "../../_firebase";

class ChatService {

    constructor() {
        this.chatThreadsCollection = firedb.collection('chat-threads');
    }

    //
    // ────────────────────────────────────────────────────────────────────────────── I ──────────
    //   :::::: P U B L I C   C R U D   M E T H O D S : :  :   :    :     :        :          :
    // ────────────────────────────────────────────────────────────────────────────────────────
    //
    createChatThread(chatThread) {
        const doc = this.chatThreadsCollection.doc();
        const docID = doc.id;
        chatThread.chatThreadID = docID;
        console.log(docID)
        return doc
            .set(chatThread)
            .then((documentSnapshot) => docID);
    }

    // todo: verify this makes sense here 
    getMyChats(userID) {
        return this.chatThreadsCollection
            .where('chatPerson1', '==', userID)
            .get()
            .then(snapshot => 
                snapshot.docs.map((docSnapshot) => 
                    docSnapshot.data()
            )            
        )
        .then(firstColDocs => 
            this.chatThreadsCollection
                .where('chatPerson2', '==', userID)
                .get()
                .then(snapshot => 
                    firstColDocs.concat(
                        snapshot.docs.map(
                            (docSnapshot) => 
                                docSnapshot.data())
                    )
                )
            )
    }

    getChatThread(chatThreadID) {
        return this.chatThreadsCollection
            .doc(chatThreadID)
            .get()
            .then(docSnapshot => docSnapshot.data());
    }

    getMessagesForChat(chatThreadID) {
        return this.chatThreadsCollection
            .doc(chatThreadID)
            .collection("chatMessages")
            .get()
            .then(snapshot => 
                snapshot.docs.map((docSnapshot) => 
                  docSnapshot.data()
                )
            );
    }

    addMessageToThread(chatMessage) {
        const doc = this.chatThreadsCollection.doc(chatMessage.chatThread).collection("chatMessages").doc();
        const docID = doc.id;
        chatMessage.chatMessageID = docID;
        console.log(chatMessage)
        return doc
            .set(chatMessage)
            .then((documentSnapshot) => docID);
    }

    // todo: test for nested collection
    // todo: decide if we want to allow this
    // todo see above, how we decide who gets to delete?...
    deleteChatThread(chatPerson1, chatPerson2, chatThreadID) {
        // const batch = firedb.batch();
        // batch.
        // return this`
        //     // .where('chatPerson1', '==',`rson1)
        //     // .where('chatPerson2', '==',`rson2)
        //     .doc(chatThreadID)
        //     // .collection("chatThreads")
        //     // .delete()
        //     .set(null)
        //     .then((documentSnapshot) => 
        //         documentSnapshot
        //     );
    }
    // ────────────────────────────────────────────────────────────────────────────────
}

export default new ChatService();
