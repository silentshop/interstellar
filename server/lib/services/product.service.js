// import firebase from 'firebase';
// import * as firebase from 'firebase/app';
// import "@firebase/firestore";
import firedb  from "../../_firebase";

class ProductService {

    constructor() {
        this.categorysCollection = firedb.collection('categories');
        this.productsCollection = firedb.collection('products');
        // this.userProductsCollection = firedb.collection('users-products');
        // myProductRef;
        // this.productCategoriesCollection = afs.collection('products-categories');
    }

    //
    // ────────────────────────────────────────────────────────────────────────────── I ──────────
    //   :::::: P U B L I C   C R U D   M E T H O D S : :  :   :    :     :        :          :
    // ────────────────────────────────────────────────────────────────────────────────────────
    //
    
    addProduct(product) {
        const doc = this.productsCollection.doc();
        const docID = doc.id;
        product.id = docID;
        const categoryID = product.categoryID;
        // todo: index vs duplication users-products / cat products
        return this.categorysCollection
            .doc(categoryID)
            .collection("products")
            .doc(docID)
            .set(product)
            .then((documentSnapshot) => docID);
    }

    /**
     */
    getAllProducts() {
        return this.productsCollection
            .get()
            .then(snapshot => 
                snapshot.docs.map((docSnapshot) => 
                    docSnapshot.data()
                )
            );
    }

    /**
     * @param  {string} productID
     */
    getProductById(productID) {
        console.log(productID)
        return this.productsCollection
            .doc(productID)
            .get()
            .then((documentSnapshot) => documentSnapshot.data());
    }


    /**
     * @param  {string} userID
    */
    getProductsByUserID(userID) {
        return this.productsCollection
            .where("productSellerID", "==", userID)
            .get()
            .then(snapshot => 
                snapshot.docs.map((docSnapshot) => 
                    docSnapshot.data()
                )
            );
        // return this.userProductsCollection
        //     .doc(userID)
        //     .collection('products')
        //     .get()
        //     .then(snapshot => 
        //         snapshot.docs.map((docSnapshot) => 
        //             docSnapshot.data()
        //         )
        //     );
    }

    updateProduct(product) {
        return this.productsCollection
                .doc(product.id)
                .update(product, { merge: true})
                .then((documentSnapshot) => documentSnapshot);
    }

    deleteProduct(productID) {
        return this.productsCollection
                .doc(productID)
                .delete()
                .then((documentSnapshot) => productID);
    }
    // ────────────────────────────────────────────────────────────────────────────────
}

// ProductService = new ProductService();
export default new ProductService();
