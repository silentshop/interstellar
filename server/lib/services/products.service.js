// import firebase from 'firebase';
// import * as firebase from 'firebase/app';
// import "@firebase/firestore";
const firedb = require("../../_firebase");

class ProductService {

    
    constructor() {
        this.productsCollection = firedb.collection('products');
        // this.productCategoriesCollection;
        // userProductsCollection;
        // myProductRef;
        // this.productCategoriesCollection = afs.collection('products-categories');
        // this.userProductsCollection = afs.collection('users-products');
    }

    //
    // ────────────────────────────────────────────────────────────────────────────── I ──────────
    //   :::::: P U B L I C   C R U D   M E T H O D S : :  :   :    :     :        :          :
    // ────────────────────────────────────────────────────────────────────────────────────────
    //

    /**
     */
    // convert to promise
    getAllProducts() {
        return this.productsCollection.get().then(snapshot => snapshot.docs.map((docSnapshot) => docSnapshot.data()));
    }

    /**
     * @returns string
     */
    getNewProductID() {
        const _newProductID = this.afs.createId();
        return _newProductID;
    }

    /**
     * @param  {string} productData
     * @returns Promise
     */
    addProduct(productData) {
        const _productData = JSON.parse(productData);
        const _docID = _productData.id;
        const _cat = _productData.productCategory;
        const batch = this.afs.firestore.batch();
        batch.set(this.productsCollection.doc(_docID).ref, _productData);
        batch.set(this.userProductsCollection.doc(this._userID).collection('products').doc(_docID).ref, _productData);
        batch.set(this.productCategoriesCollection.doc(_cat).collection('products').doc(_docID).ref, _productData);
        return batch.commit().then(() => _docID);
    }

    /**
     * @param  {string} key
     * @param  {{}} newProductData
     * @returns Promise
     */
    updateProduct(key, newProductData) {
        return this.productsCollection.doc(key).update(newProductData);
    }

    /**
     * @param  {Array<any>} pairArray
     * @returns Observable
     */
    updateProductQuantities(pairArray) {
        const batch = this.afs.firestore.batch();
        pairArray.map(pair => {
            const prodID = pair.productID;
            const sellerID = pair.sellerID;
            const newProdQuant = pair.newQuantity;
            const category = pair.category;
            const quantPairDict = {quantity: newProdQuant};
            console.log(newProdQuant);
            batch.update(this.productsCollection.doc(prodID).ref, quantPairDict);
            batch.update(this.userProductsCollection.doc(sellerID).collection('products').doc(prodID).ref, quantPairDict);
            batch.update(this.productCategoriesCollection.doc(category).collection('products').doc(prodID).ref, quantPairDict);
        });
        return batch.commit();
    }

    /**
     * @param  {string} productID
     * @param  {string} category
     * @returns Promise
     */
    deleteProduct(productID, category) {
        const batch = this.afs.firestore.batch();
        batch.delete(this.productsCollection.doc(productID).ref);
        batch.delete(this.userProductsCollection.doc(`${this._userID}/products/${productID}`).ref);
        batch.delete(this.productCategoriesCollection.doc(`${category}/products/${productID}`).ref);
        return batch.commit();
    }
    // ────────────────────────────────────────────────────────────────────────────────


    //
    // ────────────────────────────────────────────────────────────────── I ──────────
    //   :::::: Q U E R Y   M E T H O D S : :  :   :    :     :        :          :
    // ────────────────────────────────────────────────────────────────────────────
    //
    /**
     // TODO: TEST THESE and IMPLEMENT ALGOLIA!!!
     * @param  {string} productID
     * @returns Observable
     */
    getProductByProductId(productID) {
        return Observable.create((observer) => {
            this.afs.collection('products', ref => ref.where('id', '==', productID))
                .valueChanges()
                .subscribe(prod => {
                    observer.next(prod[0]);
                });
        });
    }

    getProductQuantity(productID) {
        return Observable.create((observer) => {
            // this.productsCollection
            this.afs.collection('products', ref => ref.where('id', '==', productID))
                .valueChanges()
                .subscribe(prod => {
                    // const _product = prod as Product;
                    console.log(prod);
                    observer.next(prod[0]);
                });
        });
    }

    /**
     * @param  {string} userID
     * @returns Observable
     */
    getProductsByUserID(userID) {
        return this.userProductsCollection.doc(userID).collection('products').valueChanges();
    }

    getProductsByUserName(name) {
        return;
    }

    /**
     * @param  {string} category
     * @returns Observable
     */
    getProductsByCategory(category) {
        return this.productCategoriesCollection.doc(category).collection('products').valueChanges();
    }
    // ────────────────────────────────────────────────────────────────────────────────
}

ProductService = new ProductService();
// export default ProductService;
module.exports = ProductService;
