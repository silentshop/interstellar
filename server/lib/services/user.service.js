import firedb  from "../../_firebase";
import stellar from 'stellar-sdk';

class UserService {

    constructor() {
        this.usersCollection = firedb.collection('users');
    }

    //
    // ────────────────────────────────────────────────────────────────────────────── I ──────────
    //   :::::: P U B L I C   C R U D   M E T H O D S : :  :   :    :     :        :          :
    // ────────────────────────────────────────────────────────────────────────────────────────
    //
    
    /**
     */
    getAllUsers() {
        return this.usersCollection
            .get()
            .then(snapshot => 
                snapshot.docs.map((docSnapshot) => 
                    docSnapshot.data()
                )
            );
    }

    /**
     * @param  {string} userID
     */
    getUserById(userID) {
        return this.usersCollection
            .doc(userID)
            .get()
            .then((documentSnapshot) => 
                documentSnapshot.data()
            );
    }

    createUser(user) {
        const doc = this.usersCollection.doc();
        const docID = doc.id;
        user.id = docID;
        // todo: check if better return from post
        return doc
                .set(user)
                .then((documentSnapshot) => docID);
    }

    updateUser(user) {
        return this.usersCollection
                .doc(user.id)
                .update(user, { merge: true})
                .then((documentSnapshot) => console.log(documentSnapshot));
    }

    /**
     * @param  {string} userID
     */
    // todo: test for first or take
    // todo: test for not found user
    getUserByPublicKey(publicKey) {
        return this.usersCollection
            .where("publicKey", "==", publicKey)
            .get()
            .then((QuerySnapshot) => {
                if(QuerySnapshot.docs[0])
                    return QuerySnapshot.docs[0].data();
                else
                    throw 'no user';
            });
    }

    verifySignature(publicKey, payload, signature){
        return stellar.Keypair.fromPublicKey(publicKey).verify(payload, Buffer.from(signature, 'base64'));
    }

    deleteUser(userID) {
        return this.usersCollection
                .doc(userID)
                .delete()
                .then((documentSnapshot) => userID);
    }
    // ────────────────────────────────────────────────────────────────────────────────
}

export default new UserService();
