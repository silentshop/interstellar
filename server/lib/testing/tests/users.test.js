import * as UserService from '../../services/users';
import { testUsers } from '../test-data/users.data';

// expect.addSnapshotSerializer({
//     test:(val) => val.id && val.category && val.description && val.imageURL,
//     print:(val) => 
//      `id: ${val.id}
//       category: ${val.category}
//       description: ${val.description}
//       imageURL: ${val.imageURL}`
// })

/** Test get sellers users */
test('User-Service: getAllUsers()', () => {
    UserService.getAllUsers = jest.fn().mockImplementation(() => {
        return new Promise((resolve, reject) => {
            resolve(testUsers)
        })
    });
    UserService.getAllUsers()
        .then(allUsers => 
            expect(allUsers).toMatchSnapshot()
        );
});

/** test get user by id */
test('User-Service: getUserById(userID)', () => {
    UserService.getUserById = jest.fn().mockImplementation(() => {
        return new Promise((resolve, reject) => {
            resolve({
                id: "123",
                userName: "John Doe"
            })
        })
    });
    UserService.getUserById("123")
        .then(response => 
            expect(response).toEqual({
                id: "123",
                userName: "John Doe"
            })
        );
});

/** test add new user */
test('User-Service: createUser(userData)', () => {
    UserService.createUser = jest.fn().mockImplementation(() => {
        return new Promise((resolve, reject) => {
            resolve({
                docID: "some-new-docID"
            })
        })
    });
    UserService.createUser("some-new-user-data")
        .then(response => 
            expect(response).toEqual({docID: "some-new-docID"})
        );
});

// const userId = "7KtTK9jXBT9HaU0bZ3uC";
// test('User-Service: getUserById() for user: 7KtTK9jXBT9HaU0bZ3uC', () => {
//     return Users.default.getUserById(userId).then(user => {
//         // console.log(user)
//         // expect.assertions(11)
//         expect(user).toBeDefined();
//         expect(user).toMatchObject({
//             address: expect.anything(),
//             birthdate: expect.any(String),
//             email: expect.any(String),
//             fullName: expect.any(String),
//             publicKey: expect.any(String),
//             userName: expect.any(String)
//               // expect(user.address).toMatchObject({
//             //     shipPhone: expect.any(String),
//             //     shipContactName: expect.any(String),
//             //     shipCompany_name: expect.any(String),
//             //     shipState: expect.any(String),
//             //     shipEmail: expect.any(String),
//             //     shipCountry: expect.any(String),
//             //     shipCity: expect.any(String),
//             // }),
//             // phone: expect.any(Number),
//             // numberOfItemsSold: expect.any(String),
//             // myProducts: expect.any(String),
//             // acceptedAssets: expect.any(String)
//         });
//     });
// });

// const newUser = {

// };

// test('User-Service: createUser()', () => {
//     // expect(newUser).toBeInstanceOf(User) .. ? 
// expect(user).toMatchObject({
//     address: expect.anything(),
//     birthdate: expect.any(String),
//     email: expect.any(String),
//     fullName: expect.any(String),
//     publicKey: expect.any(String),
//     userName: expect.any(String),
    
//     return Users.default.createUser().then(newUserID => {
//         console.log(newUserID)
//         return expect(newUserID).toBeDefined();
//     });
// });


// const privSignatureToVerify = "...";
// test('User-Service: verifySignature() for user: ...', () => {
//     expect(privSignatureToVerify).toBeDefined();
//     const localPubFromPriv = stellar.Keypair.fromPublicKey(privSignatureToVerify);
//     const serverPubFromPriv = Users.default.verifySignature(/* what goes here */);
//     expect(localPubFromPriv).toBeDefined();
//     expect(serverPubFromPriv).toBeDefined();
//     expect(localPubFromPriv).toEqual(serverPubFromPriv)
//     expect(localPubFromPriv).toBe(serverPubFromPriv)
// });

// const toDeleteUserId = "...";
// test('User-Service: deleteUser() for user: ...', () => {
//     return Users.default.deleteUserId(toDeleteUserId).then(deletedUserId => {
//         console.log(deletedUserId)
//         expect(deletedUserId).toBeDefined();
//         expect(deletedUserId).toBe(toDeleteUserId);
//     });
// });