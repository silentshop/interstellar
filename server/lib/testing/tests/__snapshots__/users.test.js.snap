// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User-Service: getAllUsers() 1`] = `
Array [
  Object {
    "id": "123",
    "userName": "John Doe",
  },
  Object {
    "id": "456",
    "userName": "jane doner the 3rd",
  },
]
`;
